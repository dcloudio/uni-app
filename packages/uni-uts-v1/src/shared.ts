import os from 'os'
import path from 'path'
import { parse } from 'jsonc-parser'

export function parseJson(jsonStr: string) {
  return parse(jsonStr)
}

export function once<T extends (...args: any[]) => any>(
  fn: T,
  ctx: unknown = null
): T {
  let res: any
  return ((...args: any[]) => {
    if (fn) {
      res = fn.apply(ctx, args)
      fn = null as any
    }
    return res
  }) as T
}

export const runByHBuilderX = once(() => {
  return (
    !!process.env.UNI_HBUILDERX_PLUGINS &&
    (!!process.env.RUN_BY_HBUILDERX || !!process.env.HX_Version)
  )
})

export const isInHBuilderX = once(() => {
  try {
    // eslint-disable-next-line no-restricted-globals
    const { name } = require(path.resolve(
      process.cwd(),
      '../about/package.json'
    ))
    if (name === 'about') {
      process.env.UNI_HBUILDERX_PLUGINS = path.resolve(process.cwd(), '..')
      return true
    }
  } catch (e) {
    // console.error(e)
  }
  return false
})

export function resolveSourceMapPath(
  outputDir?: string,
  platform?: UniApp.PLATFORM
) {
  let dir = platform || process.env.UNI_SUB_PLATFORM || process.env.UNI_PLATFORM
  if (dir === 'app-plus') {
    dir = 'app'
  }
  return path.resolve(
    outputDir || process.env.UNI_OUTPUT_DIR,
    '../.sourcemap/' + dir
  )
}

export const isWindows = os.platform() === 'win32'
export function normalizePath(id: string): string {
  return isWindows ? id.replace(/\\/g, '/') : id
}

function supportAutoInstallPlugin() {
  return !!process.env.HX_Version
}

export function installHBuilderXPlugin(plugin: string) {
  if (!supportAutoInstallPlugin()) {
    return
  }
  return console.error(
    `%HXRunUniAPPPluginName%${plugin}%HXRunUniAPPPluginName%`
  )
}
