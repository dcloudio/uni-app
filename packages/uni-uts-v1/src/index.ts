import { extend, isArray } from '@vue/shared'
import { extname, join, relative, resolve } from 'path'

import { resolveAndroidDepFiles } from './kotlin'
import { resolveIOSDepFiles } from './swift'

import {
  FORMATS,
  type GenProxyCodeOptions,
  genProxyCode,
  resolvePlatformIndex,
  resolvePlatformIndexFilename,
  resolveRootIndex,
} from './code'
import {
  ERR_MSG_PLACEHOLDER,
  genConfigJson,
  resolveAndroidComponents,
  resolveConfigProvider,
  resolveIOSComponents,
  resolvePackage,
} from './utils'
import { parseUTSSwiftPluginStacktrace } from './stacktrace'
import { resolveUTSPluginSourceMapFile } from './sourceMap'
import { isWindows, normalizePath } from './shared'
import {
  generateCodeFrameWithKotlinStacktrace,
  generateCodeFrameWithSwiftStacktrace,
} from './legacy'
import {
  checkCompile,
  genManifestFile,
  initCheckOptionsEnv,
  restoreDex,
  restoreSourceMap,
  storeSourceMap,
} from './manifest'
import { cacheTips } from './manifest/utils'
import { compileEncrypt, isEncrypt } from './encrypt'
import type { UTSOutputOptions } from '@dcloudio/uts'
import { isModule } from './module'
import { getCompiler } from './compiler'
import { uvueOutDir } from './uvue/index'
import {
  type SyncUniModulesFilePreprocessor,
  type UniXCompilerPlatform,
  compileUniModuleWithTsc,
  createUniXArkTSCompiler,
  createUniXKotlinCompiler,
  createUniXSwiftCompiler,
  resolveOutputPluginDir,
  resolveUVueOutputPluginDir,
} from './uni_modules'
import { existsSync, readdirSync, rmSync } from 'fs-extra'

export * from './tsc'

export { getPluginInjectApis, getPluginInjectComponents } from './utils'

export { parseExportIdentifiers } from './code'

export {
  compileArkTS,
  compileArkTSExtApi,
  getArkTSAutoImports,
  resolveAppHarmonyUniModulesRootDir,
  resolveAppHarmonyUniModulesEntryDir,
} from './arkts'

export const sourcemap = {
  generateCodeFrameWithKotlinStacktrace,
  generateCodeFrameWithSwiftStacktrace,
}

export { compileApp, CompileAppOptions } from './uvue/index'

export { parseInjectModules, parseExtApiProviders } from './utils'

export * from './sourceMap'

export {
  parseRuntimeStacktrace,
  parseUTSRuntimeStacktrace,
  parseUTSKotlinRuntimeStacktrace,
  parseUTSJavaScriptRuntimeStacktrace,
} from './stacktrace'

export { compile as toKotlin } from './kotlin'
export { compile as toSwift } from './swift'

function parseErrMsg(code: string, errMsg: string) {
  return code.replace(ERR_MSG_PLACEHOLDER, errMsg)
}

function compileErrMsg(id: string) {
  return `uts插件[${id}]编译失败，无法使用`
}

function warn(msg: string) {
  console.warn(`提示：${msg}`)
}

export interface CompileResult {
  code: string
  deps: string[]
  encrypt: boolean
  meta?: any
  dir: string
  inject_apis: string[]
  scoped_slots: string[]
}

function createResult(
  dir: string,
  errMsg: string,
  code: string,
  deps: string[],
  inject_apis: string[],
  scoped_slots: string[],
  meta: unknown
): CompileResult {
  return {
    dir,
    code: parseErrMsg(code, errMsg),
    deps,
    encrypt: false,
    inject_apis,
    scoped_slots,
    meta,
  }
}

export interface UTSPluginCompilerOptions {
  isX: boolean
  isPlugin: boolean
  isSingleThread: boolean
  isExtApi?: boolean
  extApis?: Record<string, [string, string]>
  transform?: UTSOutputOptions['transform']
  sourceMap?: boolean
  uni_modules?: string[]
  kotlinAutoImports?: () => Promise<
    Required<UTSOutputOptions>['transform']['autoImports']
  >
  swiftAutoImports?: () => Promise<
    Required<UTSOutputOptions>['transform']['autoImports']
  >
  androidPreprocessor?: SyncUniModulesFilePreprocessor
  iosPreprocessor?: SyncUniModulesFilePreprocessor
}

// 重要：当调整参数时，需要同步调整 vue2 编译器 uni-cli-shared/lib/uts/uts-loader.js
export async function compile(
  pluginDir: string,
  compilerOptions: UTSPluginCompilerOptions = {
    isX: false,
    isPlugin: true,
    isSingleThread: true,
  }
): Promise<CompileResult | void> {
  const isCompileUniModules = process.env.UNI_COMPILE_TARGET === 'uni_modules'
  const pkg = resolvePackage(pluginDir)
  if (!pkg) {
    return
  }

  // 加密插件
  if (isEncrypt(pluginDir)) {
    return compileEncrypt(pluginDir, compilerOptions.isX)
  }
  const {
    isX,
    extApis,
    isExtApi,
    transform,
    sourceMap,
    isSingleThread,
    uni_modules,
  } = compilerOptions

  let isPlugin = compilerOptions.isPlugin

  const cacheDir = process.env.HX_DEPENDENCIES_DIR || ''
  const inputDir = process.env.UNI_INPUT_DIR
  const outputDir = process.env.UNI_OUTPUT_DIR
  const utsPlatform = process.env.UNI_UTS_PLATFORM

  const pluginRelativeDir = relative(
    isCompileUniModules ? uvueOutDir('app-android') : inputDir,
    pluginDir
  )
  const outputPluginDir = normalizePath(join(outputDir, pluginRelativeDir))

  const indexModuleFilename = isModule(pluginDir)
    ? normalizePath(join(pluginDir, 'index.module.uts'))
    : ''

  if (indexModuleFilename) {
    isPlugin = false
  }

  const androidComponents = indexModuleFilename
    ? {}
    : resolveAndroidComponents(pluginDir, pkg.is_uni_modules)
  const iosComponents = indexModuleFilename
    ? {}
    : resolveIOSComponents(pluginDir, pkg.is_uni_modules)

  const env = initCheckOptionsEnv()
  const deps: string[] = []
  const inject_apis: string[] = []
  const scoped_slots: string[] = []

  const meta = {
    exports: {},
    types: {},
    typeParams: [],
    components: [],
  }
  let moduleName = pkg.id
  try {
    moduleName = require(join(pluginDir, 'package.json')).displayName
  } catch (e) {}
  const proxyCodeOptions: GenProxyCodeOptions = extend(
    {
      androidComponents,
      iosComponents,
      format:
        process.env.UNI_UTS_JS_CODE_FORMAT === 'cjs' ? FORMATS.CJS : FORMATS.ES,
      pluginRelativeDir,
      moduleName,
      moduleType: process.env.UNI_UTS_MODULE_TYPE || '',
      meta,
      inputDir,
      isExtApi,
      androidPreprocessor: compilerOptions.androidPreprocessor,
      iosPreprocessor: compilerOptions.iosPreprocessor,
    },
    pkg
  )

  const code = isCompileUniModules
    ? ''
    : await genProxyCode(pluginDir, proxyCodeOptions)

  let errMsg = ''
  if (process.env.NODE_ENV !== 'development' || isCompileUniModules) {
    // uts 插件 wgt 模式，本地资源模式不需要编译、ext-api模式也不需要编译（已经有前置编译过了）
    if (
      process.env.UNI_APP_PRODUCTION_TYPE === 'WGT' ||
      process.env.UNI_COMPILE_TARGET === 'ext-api'
    ) {
      return createResult(outputPluginDir, errMsg, code, deps, [], [], meta)
    }
    // 生产模式 支持同时生成 android 和 ios 的 uts 插件
    if (
      utsPlatform === 'app-android' ||
      utsPlatform === 'app' ||
      utsPlatform === 'app-plus'
    ) {
      let filename =
        indexModuleFilename ||
        resolvePlatformIndex('app-android', pluginDir, pkg) ||
        resolveRootIndex(pluginDir, pkg)
      if (!filename && Object.keys(androidComponents).length) {
        filename = resolvePlatformIndexFilename('app-android', pluginDir, pkg)
      }
      if (filename) {
        const result = await getCompiler('kotlin').runProd(filename, {
          components: androidComponents,
          uniModuleId: pkg.id,
          isX,
          isSingleThread,
          isPlugin,
          isExtApi,
          isModule: !!indexModuleFilename,
          extApis,
          transform: await initCompilerOptionsTransform(
            'kotlin',
            compilerOptions
          ),
          sourceMap: !!sourceMap,
          hookClass: proxyCodeOptions.androidHookClass || '',
          uniModules: uni_modules || [],
          outFilename: indexModuleFilename
            ? normalizePath(
                join(pluginRelativeDir, 'utssdk', 'app-android', 'index.kt')
              )
            : '',
        })
        if (result) {
          if (result.inject_apis) {
            inject_apis.push(...result.inject_apis)
          }
          if (result.scoped_slots) {
            scoped_slots.push(...result.scoped_slots)
          }
        }
        if (!isCompileUniModules && cacheDir) {
          // 存储 sourcemap
          storeSourceMap(
            'app-android',
            pluginRelativeDir,
            outputDir,
            cacheDir,
            pkg.is_uni_modules
          )
          genManifestFile('app-android', {
            pluginDir,
            env,
            cacheDir,
            pluginRelativeDir,
            is_uni_modules: pkg.is_uni_modules,
          })
        }
      }
    }
    if (
      utsPlatform === 'app-ios' ||
      utsPlatform === 'app' ||
      utsPlatform === 'app-plus'
    ) {
      let filename =
        indexModuleFilename ||
        resolvePlatformIndex('app-ios', pluginDir, pkg) ||
        resolveRootIndex(pluginDir, pkg)
      if (!filename && Object.keys(androidComponents).length) {
        filename = resolvePlatformIndexFilename('app-ios', pluginDir, pkg)
      }
      if (filename) {
        const result = await getCompiler('swift').runProd(filename, {
          components: iosComponents,
          uniModuleId: pkg.id,
          isX,
          isSingleThread,
          isPlugin: true, // iOS 目前仅有 plugin 模式
          isExtApi,
          extApis,
          transform: await initCompilerOptionsTransform(
            'swift',
            compilerOptions
          ),
          sourceMap: !!sourceMap,
          hookClass: proxyCodeOptions.iOSHookClass || '',
          uniModules: uni_modules || [],
        })
        if (result) {
          if (result.inject_apis) {
            inject_apis.push(...result.inject_apis)
          }
          if (result.scoped_slots) {
            scoped_slots.push(...result.scoped_slots)
          }
        }
        if (!isCompileUniModules && cacheDir) {
          storeSourceMap(
            'app-ios',
            pluginRelativeDir,
            outputDir,
            cacheDir,
            pkg.is_uni_modules
          )
          genManifestFile('app-ios', {
            pluginDir,
            env,
            cacheDir,
            pluginRelativeDir,
            is_uni_modules: pkg.is_uni_modules,
          })
        }
      }
    }
  } else {
    const compilerType = utsPlatform === 'app-android' ? 'kotlin' : 'swift'
    const versionTips = getCompiler(compilerType).checkVersionTips(
      pkg.id,
      pluginDir,
      pkg.is_uni_modules
    )
    // iOS windows 平台，标准基座不编译
    if (utsPlatform === 'app-ios') {
      if (isWindows) {
        process.env.UNI_UTS_ERRORS = `iOS手机在windows上使用标准基座真机运行无法使用uts插件，如需使用uts插件请提交云端打包自定义基座`
        return createResult(outputPluginDir, errMsg, code, deps, [], [], meta)
      }
    }
    if (utsPlatform === 'app-android' || utsPlatform === 'app-ios') {
      const components =
        utsPlatform === 'app-android' ? androidComponents : iosComponents
      let tips = ''
      // dev 模式
      if (cacheDir) {
        // 检查缓存
        // let start = Date.now()
        // console.log('uts插件[' + pkg.id + ']start', start)
        const res = await checkCompile(
          utsPlatform,
          process.env.HX_USE_BASE_TYPE,
          {
            id: pkg.id,
            env,
            cacheDir,
            outputDir,
            pluginDir,
            pluginRelativeDir,
            is_uni_modules: pkg.is_uni_modules,
          }
        )
        if (res.tips) {
          tips = res.tips
        }
        // console.log('uts插件[' + pkg.id + ']end', Date.now())
        // console.log('uts插件[' + pkg.id + ']缓存检查耗时：', Date.now() - start)
        if (!res.expired) {
          if (utsPlatform === 'app-android') {
            restoreDex(
              pluginRelativeDir,
              cacheDir,
              outputDir,
              pkg.is_uni_modules
            )
          }

          // 还原 sourcemap
          restoreSourceMap(
            utsPlatform,
            pluginRelativeDir,
            outputDir,
            cacheDir,
            pkg.is_uni_modules
          )
          // 处理 config.json
          genConfigJson(
            utsPlatform,
            isX,
            (utsPlatform === 'app-android'
              ? proxyCodeOptions.androidHookClass
              : proxyCodeOptions.iOSHookClass) || '',
            components,
            pluginRelativeDir,
            pkg.is_uni_modules,
            inputDir,
            outputDir,
            resolveConfigProvider(utsPlatform, pkg.id, transform)
          )

          console.log(cacheTips(pkg.id))

          if (res.tips) {
            warn(res.tips)
          }
          if (versionTips) {
            warn(versionTips)
          }
          // 所有文件加入依赖
          return createResult(
            outputPluginDir,
            errMsg,
            code,
            res.files.map((name) => join(pluginDir, name)),
            [],
            [],
            meta
          )
        }
      }
      let filename =
        resolvePlatformIndex(utsPlatform, pluginDir, pkg) ||
        resolveRootIndex(pluginDir, pkg)

      if (!filename && Object.keys(androidComponents).length) {
        filename = resolvePlatformIndexFilename(utsPlatform, pluginDir, pkg)
      }

      if (filename) {
        deps.push(filename)
        if (utsPlatform === 'app-android') {
          deps.push(...resolveAndroidDepFiles(filename))
        } else {
          deps.push(...resolveIOSDepFiles(filename))
        }
        // 处理 config.json
        genConfigJson(
          utsPlatform,
          isX,
          (utsPlatform === 'app-android'
            ? proxyCodeOptions.androidHookClass
            : proxyCodeOptions.iOSHookClass) || '',
          components,
          pluginRelativeDir,
          pkg.is_uni_modules,
          inputDir,
          outputDir,
          resolveConfigProvider(utsPlatform, pkg.id, transform)
        )
        const res = await getCompiler(compilerType).runDev(filename, {
          components,
          isX,
          isSingleThread,
          isPlugin,
          isExtApi,
          cacheDir,
          pluginRelativeDir,
          is_uni_modules: pkg.is_uni_modules,
          extApis,
          transform: await initCompilerOptionsTransform(
            compilerType,
            compilerOptions
          ),
          sourceMap: !!sourceMap,
          uniModules: uni_modules || [],
        })
        if (res) {
          if (isArray(res.deps) && res.deps.length) {
            // 添加其他文件的依赖
            deps.push(...res.deps)
          }
          let isSuccess = false
          if (res.type === 'swift') {
            if (res.code) {
              errMsg = compileErrMsg(pkg.id)
              try {
                console.error(
                  `error: ` +
                    (await parseUTSSwiftPluginStacktrace({
                      stacktrace: res.msg,
                      sourceMapFile: resolveUTSPluginSourceMapFile(
                        'swift',
                        filename,
                        inputDir,
                        outputDir
                      ),
                      sourceRoot: inputDir,
                    }))
                )
              } catch (e) {
                console.error(`error: ` + res.msg)
              }
            } else {
              isSuccess = true
            }
          } else if (res.type === 'kotlin') {
            if (res.changed.length) {
              isSuccess = true
            }
          }
          if (isSuccess) {
            // 生成缓存文件
            if (cacheDir) {
              // 存储 sourcemap
              storeSourceMap(
                utsPlatform,
                pluginRelativeDir,
                outputDir,
                cacheDir,
                pkg.is_uni_modules
              )
              // 生成 manifest
              genManifestFile(utsPlatform, {
                pluginDir,
                env,
                cacheDir,
                pluginRelativeDir,
                is_uni_modules: pkg.is_uni_modules,
              })
            }
            if (tips) {
              warn(tips)
            }
            if (versionTips) {
              warn(versionTips)
            }
          }
          const files: string[] = []
          if (process.env.UNI_APP_UTS_CHANGED_FILES) {
            try {
              files.push(...JSON.parse(process.env.UNI_APP_UTS_CHANGED_FILES))
            } catch (e) {}
          }
          if (res.changed && res.changed.length) {
            files.push(...res.changed)
          } else {
            if (res.type === 'kotlin') {
              errMsg = compileErrMsg(pkg.id)
            }
          }
          process.env.UNI_APP_UTS_CHANGED_FILES = JSON.stringify([
            ...new Set(files),
          ])
        } else {
          errMsg = compileErrMsg(pkg.id)
        }
      }
    }
  }
  return createResult(
    outputPluginDir,
    errMsg,
    code,
    deps,
    inject_apis,
    scoped_slots,
    meta
  )
}

export async function initCompilerOptionsTransform(
  compilerType: 'kotlin' | 'swift',
  options: UTSPluginCompilerOptions
) {
  const transform: Required<UTSPluginCompilerOptions>['transform'] = JSON.parse(
    JSON.stringify(options.transform || {})
  )
  const customAutoImports = await initCompilerOptionsTransformAutoImports(
    compilerType === 'kotlin'
      ? options.kotlinAutoImports
      : options.swiftAutoImports
  )
  if (customAutoImports && Object.keys(customAutoImports).length) {
    const autoImports = transform.autoImports || {}
    Object.keys(customAutoImports).forEach((source) => {
      if (autoImports[source]) {
        autoImports[source].push(...customAutoImports[source])
      } else {
        autoImports[source] = customAutoImports[source]
      }
    })
    transform.autoImports = autoImports
  }

  return transform
}

async function initCompilerOptionsTransformAutoImports(
  autoImports: UTSPluginCompilerOptions['kotlinAutoImports']
) {
  return autoImports?.()
}

function emptyCacheDir(
  platform: 'app-android' | 'app-ios' | 'app-harmony',
  inputDir: string,
  pluginDir: string
) {
  const outputPluginDir = resolveOutputPluginDir(platform, inputDir, pluginDir)
  const uvueOutputPluginDir = resolveUVueOutputPluginDir(
    platform,
    inputDir,
    pluginDir
  )
  if (existsSync(outputPluginDir)) {
    emptyDir(outputPluginDir)
  }
  if (existsSync(uvueOutputPluginDir)) {
    emptyDir(uvueOutputPluginDir)
  }
}

function emptyDir(dir: string) {
  try {
    for (const file of readdirSync(dir)) {
      // node >= 14.14.0
      rmSync(resolve(dir, file), { recursive: true, force: true })
    }
  } catch (e) {}
}

/**
 * 发行模式(会自动清理临时目录)
 * 目前仅 ext-api 框架的地方使用了
 * @param platform
 * @param pluginDir
 * @param param2
 * @param compilerOptions
 */
export async function buildUniModules(
  platform: 'app' | 'app-android' | 'app-ios' | 'app-harmony',
  pluginDir: string,
  options: {
    syncUniModulesFilePreprocessors: {
      android: SyncUniModulesFilePreprocessor
      ios: SyncUniModulesFilePreprocessor
      harmony: SyncUniModulesFilePreprocessor
    }
    transformVueFile?: (
      platform: UniXCompilerPlatform,
      fileName: string
    ) => Promise<string>
    rootFiles?: string[]
  },
  compilerOptions: UTSPluginCompilerOptions
) {
  const inputDir = process.env.UNI_INPUT_DIR

  const syncUniModulesFilePreprocessors = options.transformVueFile
    ? patchSyncUniModulesFilePreprocessors(
        options.syncUniModulesFilePreprocessors,
        options.transformVueFile
      )
    : options.syncUniModulesFilePreprocessors

  if (platform === 'app-android' || platform === 'app') {
    const platform = 'app-android'
    emptyCacheDir(platform, inputDir, pluginDir)
    await compileUniModuleWithTsc(
      'app-android',
      pluginDir,
      createUniXKotlinCompiler(),
      {
        rootFiles: options.rootFiles,
        preprocessor: syncUniModulesFilePreprocessors.android,
      }
    )
  }
  if (platform === 'app-ios' || platform === 'app') {
    const platform = 'app-ios'
    emptyCacheDir(platform, inputDir, pluginDir)
    await compileUniModuleWithTsc(
      'app-ios',
      pluginDir,
      createUniXSwiftCompiler(),
      {
        rootFiles: options.rootFiles,
        preprocessor: syncUniModulesFilePreprocessors.ios,
      }
    )
  }
  if (platform === 'app-harmony') {
    const platform = 'app-harmony'
    emptyCacheDir(platform, inputDir, pluginDir)
    await compileUniModuleWithTsc(
      'app-harmony',
      pluginDir,
      createUniXArkTSCompiler(),
      {
        rootFiles: options.rootFiles,
        preprocessor: syncUniModulesFilePreprocessors.harmony,
      }
    )
  }
  if (process.env.UNI_UTS_MODULE_TYPE === 'built-in') {
    compilerOptions.androidPreprocessor =
      syncUniModulesFilePreprocessors.android
    compilerOptions.iosPreprocessor = syncUniModulesFilePreprocessors.ios
  }
  return compile(pluginDir, compilerOptions)
}

function patchSyncUniModulesFilePreprocessors(
  syncUniModulesFilePreprocessors: {
    android: SyncUniModulesFilePreprocessor
    ios: SyncUniModulesFilePreprocessor
    harmony: SyncUniModulesFilePreprocessor
  },
  transformVueFile: (
    platform: UniXCompilerPlatform,
    fileName: string
  ) => Promise<string>
) {
  return {
    android: patchSyncUniModulesFilePreprocessor(
      'app-android',
      syncUniModulesFilePreprocessors.android,
      transformVueFile
    ),
    ios: patchSyncUniModulesFilePreprocessor(
      'app-ios',
      syncUniModulesFilePreprocessors.ios,
      transformVueFile
    ),
    harmony: patchSyncUniModulesFilePreprocessor(
      'app-harmony',
      syncUniModulesFilePreprocessors.harmony,
      transformVueFile
    ),
  }
}

function patchSyncUniModulesFilePreprocessor(
  platform: UniXCompilerPlatform,
  preprocessor: SyncUniModulesFilePreprocessor,
  transformVueFile: (
    platform: UniXCompilerPlatform,
    fileName: string
  ) => Promise<string>
) {
  return async (content: string, fileName: string) => {
    const fileExtname = extname(fileName)
    if (['.vue', '.uvue'].includes(fileExtname)) {
      return transformVueFile(platform, fileName)
    }
    return preprocessor(content, fileName)
  }
}

export * from './uni_modules'
