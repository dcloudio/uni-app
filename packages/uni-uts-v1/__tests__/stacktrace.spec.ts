import path from 'path'
import {
  parseUTSArkTSPluginStacktrace,
  parseUTSKotlinStacktrace,
  parseUTSSwiftPluginStacktrace,
  parseUTSSyntaxError,
} from '../src/stacktrace'
import { hbuilderFormatter } from '../src/stacktrace/kotlin'
import { normalizePath } from '../src/shared'
const stacktrace = `/uts-development-ios/dependences/buildFramework/template/xcode_ust_template/unimoduleTestUTS1/src/index.swift:3:12: error: cannot convert return expression of type 'Int' to return type 'String'
/uts-development-ios/dependences/buildFramework/template/xcode_ust_template/unimoduleTestUTS1/src/index.swift:6:12: error: cannot convert return expression of type 'Int' to return type 'String'
`

describe('uts:stacktrace', () => {
  test('parseUTSSwiftPluginStacktrace', async () => {
    const codes = await parseUTSSwiftPluginStacktrace({
      stacktrace,
      sourceMapFile: path.resolve(
        __dirname,
        './examples/sourcemap/index.swift.map'
      ),
      sourceRoot: '/Users/xxx/DCloud/test-uts',
    })
    expect(codes).toContain(
      `uni_modules/test-uts1/utssdk/app-ios/index.uts:2:10`
    )
    expect(codes).toContain(
      `uni_modules/test-uts1/utssdk/app-ios/index.uts:5:10`
    )
  })
  test('parseUTSSyntaxError', () => {
    const msg = parseUTSSyntaxError(
      `Error: 
x UTSCallback 已过时，详情查看 https://uniapp.dcloud.net.cn/plugin/uts-plugin.html#%E5%B8%B8%E8%A7%81%E6%8A%A5%E9%94%99.
    ,-[uni_modules/uts-alert/utssdk/app-android/index.uts:29:1]
29 | 	inputET:EditText
30 | 	callback:UTSCallback
    :           ^^^^^^^^^^^
31 |
Error: 
x UTSCallback 已过时，详情查看 https://uniapp.dcloud.net.cn/plugin/uts-plugin.html#%E5%B8%B8%E8%A7%81%E6%8A%A5%E9%94%99.
    ,-[uni_modules/uts-alert/utssdk/app-android/index.uts:29:1]
29 | 	inputET:EditText
30 | 	callback:UTSCallback
    :           ^^^^^^^^^^^
31 |`,
      ``
    )
    expect(msg.match(/at\s/g)?.length).toBe(2)
  })
  test('parseUTSKotlinStacktrace', async () => {
    const inputDir = `/Users/xxx/HBuilderProjects/test-x/unpackage/dist/dev/.kotlin/src`
    const sourceMapDir = path.resolve(
      __dirname,
      'examples/uts/unpackage/dist/dev/.sourcemap/app'
    )
    expect(
      await parseUTSKotlinStacktrace(
        [
          {
            type: 'warning',
            message:
              'Classpath entry points to a non-existent location: /Users/xxx/HBuilderProjects/test-x/unpackage/dist/dev/.kotlin/class',
          },
          {
            type: 'error',
            message: 'Unresolved reference: test',
            file: `${inputDir}/components/test/test.kt`,
            line: 33,
            column: 21,
          },
        ],
        {
          inputDir,
          sourceMapDir,
          replaceTabsWithSpace: true,
          format: hbuilderFormatter,
        }
      )
    ).toMatchSnapshot()
    expect(
      await parseUTSKotlinStacktrace(
        [
          {
            type: 'warning',
            message:
              'Classpath entry points to a non-existent location: /Users/xxx/HBuilderProjects/test-x/unpackage/dist/dev/.kotlin/class',
          },
          {
            type: 'error',
            message: 'Unresolved reference: test',
            file: `${inputDir}/index.kt`,
            line: 19,
            column: 39,
          },
        ],
        {
          inputDir,
          sourceMapDir,
          replaceTabsWithSpace: true,
          format: hbuilderFormatter,
        }
      )
    ).toMatchSnapshot()
  })
  test('parseUTSArkTSPluginStacktrace', async () => {
    const inputDir = normalizePath(path.resolve(__dirname, 'examples/uts'))
    const outputDir = normalizePath(
      path.resolve(__dirname, 'examples/uts/unpackage/dist/dev/app-harmony')
    )
    expect(
      await parseUTSArkTSPluginStacktrace(
        `1 ERROR: ArkTS:ERROR File: ${inputDir}/unpackage/debug/app-harmony-9ed02395/uni_modules/native-button/utssdk/app-harmony/index.ets:74:48
 Argument of type 'UniNativeViewElement | null' is not assignable to parameter of type 'UniNativeViewElement'.
  Type 'null' is not assignable to type 'UniNativeViewElement'.`,
        {
          inputDir,
          outputDir,
        }
      )
    ).toMatchSnapshot()
    expect(
      await parseUTSArkTSPluginStacktrace(
        `2 ERROR: ArkTS:ERROR File: ${inputDir}/unpackage/debug/app-harmony-9ed02395/uni_modules/native-button/utssdk/app-harmony/builder.ets:8:9
 Cannot find name 'param'. Did you mean 'params'?`,
        {
          inputDir,
          outputDir,
        }
      )
    ).toMatchSnapshot()
    expect(
      await parseUTSArkTSPluginStacktrace(
        `3 ERROR: ArkTS:ERROR File: ${inputDir}/unpackage/debug/app-harmony-9ed02395/uni_modules/native-button/utssdk/app-harmony/index.ets:73:17
 Cannot find name 'nativeViewElemen'. Did you mean 'nativeViewElement'?`,
        {
          inputDir,
          outputDir,
        }
      )
    ).toMatchSnapshot()
    expect(
      await parseUTSArkTSPluginStacktrace(
        `1 WARN: ArkTS:WARN File: D:/demo-projects/demo-app-x/unpackage/debug/app-harmony-fde54b25/oh_modules/.ohpm/@dcloudio+uni-app-x-runtime@8kwvevyo0ovlx9ebtjtkdkzydpkx+j5qpbuuwz5vjge=/oh_modules/@dcloudio/uni-app-x-runtime/src/main/ets/runtime/dom/NamedNodeMap.ets:16:45`,
        {
          inputDir: 'D:/demo-projects/demo-app-x/',
          outputDir:
            'D:/demo-projects/demo-app-x/unpackage/dist/dev/app-harmony',
        }
      )
    ).toMatchSnapshot()
  })
})
