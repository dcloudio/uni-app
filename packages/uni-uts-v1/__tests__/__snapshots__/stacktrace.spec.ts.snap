// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`uts:stacktrace parseUTSArkTSPluginStacktrace 1`] = `
" Argument of type 'UniNativeViewElement | null' is not assignable to parameter of type 'UniNativeViewElement'.
  Type 'null' is not assignable to type 'UniNativeViewElement'.
at uni_modules/native-button/utssdk/app-harmony/index.uts:91:47
89 |              const nativeViewElement = iterateElement(ins.$el as UniElement)
90 |              if (nativeViewElemen != null) {
91 |                  return new NativeButtonContext(nativeViewElement)
   |                                                 ^
92 |              }
93 |          }"
`;

exports[`uts:stacktrace parseUTSArkTSPluginStacktrace 2`] = `
" Cannot find name 'param'. Did you mean 'params'?
at uni_modules/native-button/utssdk/app-harmony/builder.ets:8:9
6  |      .onClick(() => {
7  |          console.log('params 123',params)
8  |          param.onClick()
   |           ^
9  |      })
10 |      .attributeModifier(params.modifier)"
`;

exports[`uts:stacktrace parseUTSArkTSPluginStacktrace 3`] = `
" Cannot find name 'nativeViewElemen'. Did you mean 'nativeViewElement'?
at uni_modules/native-button/utssdk/app-harmony/index.uts:90:16
88 |          if (ins.$el != null) {
89 |              const nativeViewElement = iterateElement(ins.$el as UniElement)
90 |              if (nativeViewElemen != null) {
   |                  ^
91 |                  return new NativeButtonContext(nativeViewElement)
92 |              }"
`;

exports[`uts:stacktrace parseUTSKotlinStacktrace 1`] = `
"‌error: Unresolved reference: test‌
at components/test/test.vue:14:28
12 |          methods: {
13 |              onClick() {
14 |                  console.log(test)
   |                              ^
15 |              }
16 |          }"
`;

exports[`uts:stacktrace parseUTSKotlinStacktrace 2`] = `
"‌error: Unresolved reference: test‌
at App.vue:4:38
2  |      export default {
3  |          onLaunch: function() {
4  |              console.log('App Launch', test)
   |                                        ^
5  |          },
6  |          onShow: function() {"
`;
